### PR TITLE
Organic queries contenders selection : prioritising best contenders for a given task

### PR DESCRIPTION
**Pull request by MahdiPresario001**
for each organic query:

- recover valid contenders for that task ordered by normalised net scores
- remove bottom 25%. Top 25% have higher weight than the rest (by a factor F) - all these params can be modified from generic_constants
- select top_x contenders from the list above randomly using the weights as probabilities
- in case the number of contenders < top_x, we do not remove the bottom 25%. we return all contenders
- in case this method returns no contenders, fall back to synth queries method (happens when the stats table has no data)

### CODE REVIEW

File: validator/db/src/sql/contenders.py [hunk 1]
```diff
@@ -98,59 +102,144 @@ async def migrate_contenders_to_contender_history(connection: Connection) -> Non
...
+            SELECT *
+            FROM ranked_contenders
             """,
             task,
-            top_x - len(rows) if rows else top_x,
-            len(rows) if rows else 0,
         )
-        rows = rows + additional_rows if rows else additional_rows
+    logger.debug(f"Number of valid contenders for task {task} for organic query : {len(rows)}")
+    rows_contenders = [
```
Comment (namoray) [inline validator/db/src/sql/contenders.py:186]: What is the point of this, cant you just use the row object? If not, at least call the keys with the same constant variable

File: validator/db/src/sql/contenders.py [hunk 2]
```diff
@@ -98,59 +102,144 @@ async def migrate_contenders_to_contender_history(connection: Connection) -> Non
...
+            "total_requests_made": row[dcst.TOTAL_REQUESTS_MADE],
+            "requests_429": row[dcst.REQUESTS_429],
+            "requests_500": row[dcst.REQUESTS_500],
+            "period_score": row.get(dcst.PERIOD_SCORE, None),
+        }
+        for row in rows
+    ]
+
+    # sort contenders by net normalised scores
+    contenders = [Contender(**row) for row in rows_contenders]
```
Comment (namoray) [inline validator/db/src/sql/contenders.py:205]: Quicker to do this in the sql query no?

File: validator/db/src/sql/contenders.py [hunk 3]
```diff
@@ -98,59 +102,144 @@ async def migrate_contenders_to_contender_history(connection: Connection) -> Non
...
+            combined_contenders = [contender_score[0] for contender_score in (best_top_x_percent + remaining_top_x_percent)]
+            combined_weights = best_top_x_weights + remaining_top_x_weights
+            
+            selected_contenders = random.choices(combined_contenders, weights=combined_weights, k=min(top_x, len(combined_contenders)))
+            logger.debug(f"Selected contenders for task {task} : {selected_contenders}")
+            return selected_contenders
+        else:
+            logger.debug(f"Number of contenders ({len(contenders_with_scores)}) < top_x ({top_x}). Returning all contenders")
+            return [contender_score[0] for contender_score in contenders_with_scores]
+    # fall back in case of an issue
```
Comment (namoray) [inline validator/db/src/sql/contenders.py:228]: obvious as that what 'else' does

File: validator/utils/generic/generic_constants.py [hunk 4]
```diff
@@ -6,3 +6,8 @@
 ERROR_MESSAGE="error_message"
 JOB_ID="job_id"
 CONTENT="content"
+
+# organic contenders selection parameters
+ORGANIC_SELECT_CONTENDER_LOW_POURC=0.75
```
Comment (namoray) [inline validator/utils/generic/generic_constants.py:11]: i think POURC is french, but yolo i like it

File: validator/db/src/sql/contenders.py (hunk 1 above)
Comment (MahdiPresario001) [reply validator/db/src/sql/contenders.py:186]: we have an additional attribute (normalised net score) so we cannot use rows directly.
making the change for the use of constant

File: validator/utils/generic/generic_constants.py (hunk 4 above)
Comment (MahdiPresario001) [reply validator/utils/generic/generic_constants.py:11]: hahaha sorry force of habit

File: validator/db/src/sql/contenders.py (hunk 2 above)
Comment (MahdiPresario001) [reply validator/db/src/sql/contenders.py:205]: yeah true! not by much i would say, but i'll change it

**Pull request merged**